auth_manager.go

```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
 	b.asyncWrites = config.Global().UseAsyncSessionWrite
 	b.store = store
 	b.store.Connect()
+
+	if b.asyncWrites {
+		// check pool size in config and set to 50 if unset
+		b.poolSize = config.Global().SessionUpdatePoolSize
+		if b.poolSize <= 0 {
+			b.poolSize = 50
+		}
+		//check size for channel buffer and set to 1000 if unset
+		b.bufferSize = config.Global().SessionUpdateBufferSize
+		if b.bufferSize <= 0 {
+			b.bufferSize = 1000
+		}
+
+		log.WithField("Auth Manager", b.poolSize).Debug("Session update async pool size")
+
+		b.updateChan = make(chan *SessionUpdate, b.bufferSize)
+
+		b.keyPrefix = b.store.GetKeyPrefix()
+
+		//start worker pool
+		atomic.SwapUint32(&b.shouldStop, 0)
+		for i := 0; i < b.poolSize; i++ {
+			b.poolWG.Add(1)
+			go b.updateWorker()
+		}
+	}
+}
+
+func (b *DefaultSessionManager) updateWorker() {
+	defer b.poolWG.Done()
+
+	for range b.updateChan {
+		// grab update object from channel
+		u := <-b.updateChan
+
+		v, err := json.Marshal(u.session)
+		if err != nil {
+			log.Error("Error marshalling session for async session update")
+			continue
+		}
+
+		if u.isHashed {
+			u.keyVal = b.keyPrefix + u.keyVal
+			err := b.store.SetRawKey(u.keyVal, string(v), u.ttl)
+			if err != nil {
+				log.Errorf("Error updating hashed key: %v", err)
```

Comment:
logrus offers api for outputting errors as a special field `log.WithError(err).Error("Error updating hashed key")`